code_annotations/find_annotations.py

```diff
@@ -0,0 +1,253 @@
+"""
+Click command to do static annotation searching via Stevedore plugins.
+"""
+import datetime
+import errno
+import os
+
+import click
+import yaml
+from stevedore import named
+
+from code_annotations.exceptions import ConfigurationException
+from code_annotations.helpers import VerboseEcho
+
+# Global logger for this script, shared with extensions
+ECHO = VerboseEcho()
+
+
+def load_failed_handler(*args, **kwargs):
+    """
+    Handle failures to load an extension.
+
+    Dumps the error and raises an exception. By default these
+    errors just fail silently.
+
+    :param args:
+    :param kwargs:
+    :raises: ConfigurationException
+    """
+    ECHO.echo(args)
+    ECHO.echo(kwargs)
+    raise Exception('Failed to load a plugin, aborting.')
+
+
+def search_extension(ext, file_handle, file_extensions_map, filename_extension):
+    """
+    Execute a search on the given file using the given extension.
+
+    :param ext:
+    :param file_handle:
+    :param file_extensions_map:
+    :param filename_extension:
+    :return: Tuple of (extension name, list of found annotation dicts)
+    """
+    # Reset the read handle to the beginning of the file in case another
+    # extension already moved it
+    file_handle.seek(0)
+
+    # Only search this file if we are configured for its extension
+    if filename_extension in file_extensions_map[ext.name]:
+        ext_results = ext.obj.search(file_handle)
+
+        if ext_results:
+            return ext.name, ext_results
+
+    return ext.name, None
+
+
+def format_file_results(source_path, all_results, results):
+    """
+    Add all extensions' search results for a file to the overall results.
+
+    :param source_path: Top level search path
+    :param all_results: Aggregated results to add the results to
+    :param results: Results of search() on a single file
+    :return: None, modifies all_results
+    """
+    # _ here is the extension name, as required by Stevedore map(). Each
+    # annotation already has the extension name so we can ignore it here
+    for _, annotations in results:
+        if annotations is None:
+            continue
+
+        # TODO: The file_path should be the same for all of these results
+        # so we should be able to optimize getting file_path and making
+        # sure it exists in the dict to do this less often.
+
+        # source_path is unnecessary here, we only want the file name
+        # relative to it.
+        file_path = annotations[0]['filename'].replace(source_path, '')
```

Comment:
Valid user input without a trailing slash can cause this relative path to become a non-existent absolute path with a leading slash.  Can you take advantage of a helper function from `pathlib` or something to do path magic?